### PR TITLE
Feature/oauth authentification

### PR DESCRIPTION
**Pull request by pwitusik**

### CODE REVIEW

File: G1ANT.Addon.Net/API/ImapManager.cs [hunk 1]
```diff
@@ -18,10 +18,7 @@ public sealed class ImapManager
         private void ConnectClient(ImapClient client)
         {
             client.Connect(uri);
-            if (credentials != null)
-                client.Authenticate(credentials);
-            if (!client.IsConnected || !client.IsAuthenticated)
-                throw new Exception("Could not connect or authenticate on the server");
+            authenticator?.Authenticate(client);
```
Comment (ChrisPrusik) [inline G1ANT.Addon.Net/API/ImapManager.cs:21]: Maybe we should throw exception when authenticator is null.

File: G1ANT.Addon.Net/API/SmtpManager.cs [hunk 2]
```diff
@@ -22,23 +19,21 @@ public sealed class SmtpManager
         private void ConnectClient(SmtpClient client)
         {
             client.Connect(host, port, socketOptions);
-            if (credentials != null)
-                client.Authenticate(credentials);
+            authenticator?.Authenticate(client);
```
Comment (ChrisPrusik) [inline G1ANT.Addon.Net/API/SmtpManager.cs:22]: Maybe we should throw exception when authenticator is null.

File: G1ANT.Addon.Net/Commands/ImapOpenCommand.cs [hunk 3]
```diff
@@ -47,9 +48,7 @@ public void Execute(Arguments arguments)
             {
                 ServicePointManager.ServerCertificateValidationCallback += (sender, cert, chain, sslPolicyErrors) => true;
             }
-            NetworkCredential credentials = null;
-            if (!string.IsNullOrEmpty(arguments.Login?.Value) && !string.IsNullOrEmpty(arguments.Password?.Value))
-                credentials = new NetworkCredential(arguments.Login.Value, arguments.Password.Value);
+            var credentials = new SimpleAuthenticationModel(arguments.Login.Value, arguments.Password.Value);
```
Comment (ChrisPrusik) [inline G1ANT.Addon.Net/Commands/ImapOpenCommand.cs:51]: In my opinion we should check all neccessary arguments and give to the RPA Developer information if there is something missing. The same in url.

File: G1ANT.Addon.Net/Commands/ImapOpenExCommand.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        public ImapOpenExCommand(AbstractScripter scripter) : base(scripter)
+        { }
+
+        public void Execute(Arguments arguments)
+        {
+            if (arguments.IgnoreCertificateErrors.Value)
+            {
+                ServicePointManager.ServerCertificateValidationCallback += (sender, cert, chain, sslPolicyErrors) => true;
+            }
+            var uri = new UriBuilder(arguments.UseSsl.Value ? "imaps" : "imap", arguments.Host.Value, arguments.Port.Value).Uri;
```
Comment (ChrisPrusik) [inline G1ANT.Addon.Net/Commands/ImapOpenExCommand.cs:47]: In my opinion we should check all neccessary arguments and give to the RPA Developer information if there is something missing. The same in url.

File: G1ANT.Addon.Net/Commands/ImapOpenExCommand.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            var timeout = (int)arguments.Timeout.Value.TotalMilliseconds;
+
+            IAuthenticationModel authenticator = null;
+            if (arguments.Authentication != null)
+            {
+                if (arguments.Authentication.Object is IAuthenticationModel model)
+                    authenticator = model;
+                else
+                    throw new ArgumentException($"Authentication argument is incorrect type, try 'simpleauthentication' structure");
```
Comment (ChrisPrusik) [inline G1ANT.Addon.Net/Commands/ImapOpenExCommand.cs:56]: Cleaner code:

```
if (arguments.Authentication.Object is not IAuthenticationModel model)
   throw new ArgumentException($"Authentication argument is incorrect type, try 'simpleauthentication' structure");    

authenticator = model;

```

File: G1ANT.Addon.Net/Commands/MailImapCommand.cs [hunk 6]
```diff
@@ -68,14 +67,12 @@ public void Execute(Arguments arguments)
             {
                 ServicePointManager.ServerCertificateValidationCallback += (sender, cert, chain, sslPolicyErrors) => true;
             }
-            NetworkCredential credentials = null;
-            if (!string.IsNullOrEmpty(arguments.Login?.Value) && !string.IsNullOrEmpty(arguments.Password?.Value))
-                credentials = new NetworkCredential(arguments.Login.Value, arguments.Password.Value);
+            var authenticator = new SimpleAuthenticationModel(arguments.Login?.Value, arguments.Password?.Value);
```
Comment (ChrisPrusik) [inline G1ANT.Addon.Net/Commands/MailImapCommand.cs:70]: In my opinion we should check all neccessary arguments and give to the RPA Developer information if there is something missing. The same in url.

File: G1ANT.Addon.Net/Commands/SmtpOpenCommand.cs [hunk 7]
```diff
@@ -46,23 +47,9 @@ public void Execute(Arguments arguments)
             {
                 ServicePointManager.ServerCertificateValidationCallback += (sender, cert, chain, sslPolicyErrors) => true;
             }
-            NetworkCredential credentials = null;
-            if (!string.IsNullOrEmpty(arguments.Login?.Value) && !string.IsNullOrEmpty(arguments.Password?.Value))
-                credentials = new NetworkCredential(arguments.Login.Value, arguments.Password.Value);
+            var credentials = new SimpleAuthenticationModel(arguments.Login.Value, arguments.Password.Value);
```
Comment (ChrisPrusik) [inline G1ANT.Addon.Net/Commands/SmtpOpenCommand.cs:50]: In my opinion we should check all neccessary arguments and give to the RPA Developer information if there is something missing.

File: G1ANT.Addon.Net/Commands/SmtpOpenExCommand.cs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        public SmtpOpenExCommand(AbstractScripter scripter) : base(scripter)
+        { }
+
+        public void Execute(Arguments arguments)
+        {
+            if (arguments.IgnoreCertificateErrors.Value)
+            {
+                ServicePointManager.ServerCertificateValidationCallback += (sender, cert, chain, sslPolicyErrors) => true;
+            }
+            var timeout = (int)arguments.Timeout.Value.TotalMilliseconds;
```
Comment (ChrisPrusik) [inline G1ANT.Addon.Net/Commands/SmtpOpenExCommand.cs:48]: In my opinion we should check all neccessary arguments and give to the RPA Developer information if there is something missing.

File: G1ANT.Addon.Net/Commands/SmtpOpenExCommand.cs [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+            IAuthenticationModel authenticator = null;
+            if (arguments.Authentication != null)
+            {
+                if (arguments.Authentication.Object is IAuthenticationModel model)
```
Comment (ChrisPrusik) [inline G1ANT.Addon.Net/Commands/SmtpOpenExCommand.cs:52]: Cleaner code:

```
if (arguments.Authentication.Object is not IAuthenticationModel model)
   throw new ArgumentException($"Authentication argument is incorrect type, try 'simpleauthentication' structure");    

authenticator = model;
```

File: G1ANT.Addon.Net/Models/OfficeOAuthModel.cs [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+using System.Threading.Tasks;
+using System.Linq;
+using G1ANT.Addon.Net.API;
+using G1ANT.Language;
+
+namespace G1ANT.Addon.Net.Models
+{
+    public class OfficeOAuthModel : IAuthenticationModel
+    {
+        public string Username { get; set; } = "";
```
Comment (ChrisPrusik) [inline G1ANT.Addon.Net/Models/OfficeOAuthModel.cs]: Why not = string.Empty? or why not null?

File: G1ANT.Addon.Net/Models/SimpleAuthenticationModel.cs [hunk 11]
```diff
@@ -0,0 +1,39 @@
+﻿using MailKit.Net.Imap;
+using MailKit.Net.Smtp;
+using System.Net;
+
+namespace G1ANT.Addon.Net.Models
+{
+    public class SimpleAuthenticationModel : IAuthenticationModel
+    {
+        public string Username { get; set; } = "";
```
Comment (ChrisPrusik) [inline G1ANT.Addon.Net/Models/SimpleAuthenticationModel.cs]: Why not = string.Empty? or why not null?

File: G1ANT.Addon.Net/Models/SimpleAuthenticationModel.cs [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+            Password = password;
+        }
+
+        public void Authenticate(ImapClient client)
+        {
+            if (!string.IsNullOrEmpty(Username) && !string.IsNullOrEmpty(Password))
+            {
+                var credentials = new NetworkCredential(Username, Password);
+                client.Authenticate(credentials);
+            }
```
Comment (ChrisPrusik) [inline G1ANT.Addon.Net/Models/SimpleAuthenticationModel.cs:27]: Is it ok if username or password is empty? or we should throw an exception?

File: G1ANT.Addon.Net/Models/SimpleAuthenticationModel.cs [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+        }
+
+        public void Authenticate(SmtpClient client)
+        {
+            if (!string.IsNullOrEmpty(Username) && !string.IsNullOrEmpty(Password))
+            {
+                var credentials = new NetworkCredential(Username, Password);
+                client.Authenticate(credentials);
+            }
+        }
```
Comment (ChrisPrusik) [inline G1ANT.Addon.Net/Models/SimpleAuthenticationModel.cs:37]: Is it ok if username or password is empty? or we should throw an exception?

File: G1ANT.Addon.Net/Structures/SimpleAuthenticationStructure.cs [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+        }
+        public override Structure Get(string index = "")
+        {
+            if (string.IsNullOrWhiteSpace(index))
+                throw new ArgumentException($"Index cannot be empty, possible values: {string.Join(", ", index)}", nameof(index));
+
+            switch (index.ToLower())
+            {
+                case IndexNames.Username:
+                    return new TextStructure(Value.Username);
```
Comment (ChrisPrusik) [inline G1ANT.Addon.Net/Structures/SimpleAuthenticationStructure.cs]: What if Value is null? Maybe Value?.Username

File: G1ANT.Addon.Net/Structures/SimpleAuthenticationStructure.cs [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+                case IndexNames.Password:
+                    return new TextStructure(Value.Password);
```
Comment (ChrisPrusik) [inline G1ANT.Addon.Net/Structures/SimpleAuthenticationStructure.cs]: What if Value is null? Maybe Value?.Password

File: G1ANT.Addon.Net/Structures/SimpleAuthenticationStructure.cs [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+
+        public override void Set(Structure structure, string index = null)
+        {
+            if (string.IsNullOrWhiteSpace(index))
+                throw new ArgumentException($"Index cannot be empty, possible values: {string.Join(", ", index)}", nameof(index));
+
+            switch (index.ToLower())
+            {
+                case IndexNames.Username:
+                    Value.Username = structure.ToString();
```
Comment (ChrisPrusik) [inline G1ANT.Addon.Net/Structures/SimpleAuthenticationStructure.cs]: maybe: structure?.ToString()

File: G1ANT.Addon.Net/Structures/SimpleAuthenticationStructure.cs [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+                    break;
+                case IndexNames.Password:
+                    Value.Password = structure.ToString();
```
Comment (ChrisPrusik) [inline G1ANT.Addon.Net/Structures/SimpleAuthenticationStructure.cs]: maybe: structure?.ToString()

**Review by ChrisPrusik — APPROVED** [review]

File: G1ANT.Addon.Net/API/ImapManager.cs (hunk 1 above)
Comment (pwitusik) [reply G1ANT.Addon.Net/API/ImapManager.cs:21]: it is possible to not pass authenticator to imap.openex/smtp.openex, this is to allow to connect to the server which doesn't need authentication

File: G1ANT.Addon.Net/API/SmtpManager.cs (hunk 2 above)
Comment (pwitusik) [reply G1ANT.Addon.Net/API/SmtpManager.cs:22]: it is possible to not pass authenticator to imap.openex/smtp.openex, this is to allow to connect to the server which doesn't need authentication

File: G1ANT.Addon.Net/Commands/ImapOpenExCommand.cs (hunk 5 above)
Comment (pwitusik) [reply G1ANT.Addon.Net/Commands/ImapOpenExCommand.cs:56]: this check should be done only when Authentication argument is passed not every time

File: G1ANT.Addon.Net/Models/SimpleAuthenticationModel.cs (hunk 12 above)
Comment (pwitusik) [reply G1ANT.Addon.Net/Models/SimpleAuthenticationModel.cs:27]: it is ok in some cases when server doesn't need user/password (Rhenus case)

File: G1ANT.Addon.Net/Models/SimpleAuthenticationModel.cs (hunk 13 above)
Comment (pwitusik) [reply G1ANT.Addon.Net/Models/SimpleAuthenticationModel.cs:37]: it is ok in some cases when server doesn't need user/password (Rhenus case)

**Pull request merged**